org.opentosca.bus.management.service.impl/src/org/opentosca/bus/management/service/impl/collaboration/DeploymentDistributionDecisionMaker.java

```diff
@@ -0,0 +1,420 @@
+        DeploymentDistributionDecisionMaker.LOG.debug("Looking for infrastructure NodeTemplate at NodeTemplate {} and below...",
+                                                      nodeTemplateInstance.getTemplateId());
+
+        final Collection<RelationshipTemplateInstance> outgoingRelationships =
+            nodeTemplateInstance.getOutgoingRelations();
+
+        // terminate search if bottom NodeTemplate is found
+        if (outgoingRelationships.isEmpty()) {
+            DeploymentDistributionDecisionMaker.LOG.debug("NodeTemplate {} is the infrastructure NodeTemplate",
+                                                          nodeTemplateInstance.getTemplateId());
+            return nodeTemplateInstance;
+        } else {
+            DeploymentDistributionDecisionMaker.LOG.debug("NodeTemplate {} has outgoing RelationshipTemplates...",
+                                                          nodeTemplateInstance.getTemplateId());
+
+            for (final RelationshipTemplateInstance relation : outgoingRelationships) {
+                final QName relationType = relation.getTemplateType();
+                DeploymentDistributionDecisionMaker.LOG.debug("Found outgoing RelationshipTemplate of type: {}",
+                                                              relationType);
+
+                // traverse topology stack downwards
+                if (isInfrastructureRelationshipType(relationType)) {
+                    DeploymentDistributionDecisionMaker.LOG.debug("Continue search with the target of the RelationshipTemplate...");
+                    return searchInfrastructureNode(relation.getTarget());
+                } else {
+                    DeploymentDistributionDecisionMaker.LOG.debug("RelationshipType is not valid for infrastructure search (e.g. hostedOn).");
+                }
+            }
+        }
+
+        // if all outgoing relationships are not of the searched types, the NodeTemplate is the
+        // bottom one
+        return nodeTemplateInstance;
+    }
+
+    /**
+     * Match the given NodeType and properties against instance data from the local repository. The
+     * matching is successful if a NodeTemplateInstance with the same NodeType and the same values
+     * for the properties is found in the instance data.
+     *
+     * @param infrastructureNodeType the NodeType of the NodeTemplate which has to be matched
+     * @param infrastructureProperties the set of properties of the NodeTemplate which has to be
+     *        matched
+     * @return the deployment location if a matching NodeTemplateInstance is found, <tt>null</tt>
+     *         otherwise.
+     */
+    protected static String performInstanceDataMatching(final QName infrastructureNodeType,
+                                                        final Map<String, String> infrastructureProperties) {
+
+        if (infrastructureNodeType != null) {
```

Comment:
this is a check of the argument. Annotate with `@NonNull` and let static code-analysis handle this